### ISSUE TITLE
could not create connection to server

### ISSUE THREAD

**Issue by SaltedSlark** [clangd/coc-clangd#577]
- paltform: macos
- coc-settings: 
```shell
  "clangd.path":"~/.config/coc/extensions/coc-clangd-data/install/15.0.6/clangd_15.0.6/bin/clangd"
```
- clangd was install in offical github-repo `clangd-mac-15.6.0.zip`

- shell cmd execute
```shell
 ~/.config/coc/extensions/coc-clangd-data/install/15.0.6/clangd_15.0.6/bin/clangd
clangd is a language server that provides IDE-like features to editors.

It should be used via an editor plugin rather than invoked directly. For more information, see:
	https://clangd.llvm.org/
	https://microsoft.github.io/language-server-protocol/

clangd accepts flags on the commandline, and in the CLANGD_FLAGS environment variable.

I[23:31:51.975] clangd version 15.0.6 (https://github.com/llvm/llvm-project 088f33605d8a61ff519c580a71b1dd57d16a03f8)
I[23:31:51.976] Features: mac+grpc+xpc
I[23:31:51.976] PID: 15297
I[23:31:51.976] Working directory: /Users/zhangjiang/workspace/faiss
I[23:31:51.976] argv[0]: /Users/zhangjiang/.config/coc/extensions/coc-clangd-data/install/15.0.6/clangd_15.0.6/bin/clangd
I[23:31:51.977] Starting LSP over stdin/stdout
```
seems okay.

I don't know what's wrong, need help THX!

**Comment by fannheyward** [comment]
`:CocOpenLog` please.

**Comment by SaltedSlark** [comment]
```shell
2023-04-27T23:39:21.461 INFO (pid:15448) [plugin] - coc.nvim initialized with node: v18.13.0 after [33m119[39m
2023-04-27T23:39:21.467 INFO (pid:15448) [services] - LanguageClient clangd state change: stopped => starting
2023-04-27T23:39:21.468 INFO (pid:15448) [language-client-index] - Language server "clangd" started with 15490
2023-04-27T23:39:21.474 INFO (pid:15448) [services] - LanguageClient clangd state change: starting => stopped
2023-04-27T23:39:21.474 INFO (pid:15448) [services] - LanguageClient clangd state change: stopped => starting
2023-04-27T23:39:21.475 ERROR (pid:15448) [language-client-client] - Server clangd initialization failed. jh [Error]: Pending response rejected since connection got disposed
```

**Comment by fannheyward** [comment]
> Server clangd initialization failed

Looks like the clangd server failed to start, then client failed to create connection. Have no idea about the server fail to start.

**Comment by fannheyward** [comment]
@oren0e `:CocOpenLog` please.

**Comment by oren0e** [comment]
@fannheyward sure, here it is: (I try to open a c file):
```
2023-05-11T11:06:56.445 WARN (pid:45126) [workspace] - Extension "coc-lists" registered synchronized autocmd "VimLeavePre", which could be slow.
2023-05-11T11:06:56.570 INFO (pid:45126) [plugin] - coc.nvim initialized with node: v19.8.0 after [33m335[39m
2023-05-11T11:07:06.899 INFO (pid:45126) [attach] - receive notification: highlight []
2023-05-11T11:07:06.925 INFO (pid:45126) [services] - LanguageClient clangd state change: stopped => starting
2023-05-11T11:07:06.941 INFO (pid:45126) [language-client-index] - Language server "clangd" started with 45173
2023-05-11T11:07:06.972 ERROR (pid:45126) [language-client] - Sending request failed.
2023-05-11T11:07:06.977 INFO (pid:45126) [services] - LanguageClient clangd state change: starting => stopped
2023-05-11T11:07:06.978 INFO (pid:45126) [services] - LanguageClient clangd state change: stopped => starting
2023-05-11T11:07:06.979 ERROR (pid:45126) [language-client-client] - Server clangd initialization failed. jh [Error]: Pending response rejected since connection got disposed
    at Object.dispose (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:51:4592)
    at Object.dispose (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:241:6209)
    at Nd.handleConnectionClosed (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:244:15575)
    at Nd.handleConnectionClosed (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:248:3561)
    at t (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:244:15224)
    at yT.invoke (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:14460)
    at lc.fire (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:15230)
    at Qe (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:25:6504)
    at yT.invoke (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:14460)
    at lc.fire (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:15230) {
  code: [33m-32097[39m,
  data: [90mundefined[39m
}
2023-05-11T11:07:06.985 INFO (pid:45126) [services] - LanguageClient clangd state change: starting => stopped
2023-05-11T11:07:07.121 INFO (pid:45126) [language-client-index] - Language server "clangd" started with 45174
2023-05-11T11:07:07.154 INFO (pid:45126) [services] - LanguageClient clangd state change: stopped => starting
2023-05-11T11:07:07.155 ERROR (pid:45126) [language-client-client] - Server clangd initialization failed. jh [Error]: Pending response rejected since connection got disposed
    at Object.dispose (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:51:4592)
    at Object.dispose (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:241:6209)
    at Nd.handleConnectionClosed (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:244:15575)
    at Nd.handleConnectionClosed (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:248:3561)
    at t (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:244:15224)
    at yT.invoke (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:14460)
    at lc.fire (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:15230)
    at Qe (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:25:6504)
    at yT.invoke (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:14460)
    at lc.fire (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:15230) {
  code: [33m-32097[39m,
  data: [90mundefined[39m
}
2023-05-11T11:07:07.156 INFO (pid:45126) [services] - LanguageClient clangd state change: starting => stopped
2023-05-11T11:07:07.156 ERROR (pid:45126) [services] - Server clangd failed to start: jh [Error]: Pending response rejected since connection got disposed
    at Object.dispose (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:51:4592)
    at Object.dispose (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:241:6209)
    at Nd.handleConnectionClosed (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:244:15575)
    at Nd.handleConnectionClosed (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:248:3561)
    at t (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:244:15224)
    at yT.invoke (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:14460)
    at lc.fire (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:15230)
    at Qe (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:25:6504)
    at yT.invoke (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:14460)
    at lc.fire (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:15230) {
  code: [33m-32097[39m,
  data: [90mundefined[39m
}
2023-05-11T11:07:07.199 INFO (pid:45126) [language-client-index] - Language server "clangd" started with 45176
2023-05-11T11:07:07.233 INFO (pid:45126) [services] - LanguageClient clangd state change: stopped => starting
2023-05-11T11:07:07.233 ERROR (pid:45126) [language-client-client] - Server clangd initialization failed. jh [Error]: Pending response rejected since connection got disposed
    at Object.dispose (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:51:4592)
    at Object.dispose (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:241:6209)
    at Nd.handleConnectionClosed (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:244:15575)
    at Nd.handleConnectionClosed (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:248:3561)
    at t (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:244:15224)
    at yT.invoke (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:14460)
    at lc.fire (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:15230)
    at Qe (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:25:6504)
    at yT.invoke (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:14460)
    at lc.fire (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:15230) {
  code: [33m-32097[39m,
  data: [90mundefined[39m
}
2023-05-11T11:07:07.234 INFO (pid:45126) [services] - LanguageClient clangd state change: starting => stopped
2023-05-11T11:07:07.245 INFO (pid:45126) [language-client-index] - Language server "clangd" started with 45177
2023-05-11T11:07:07.264 ERROR (pid:45126) [language-client-client] - Server clangd initialization failed. jh [Error]: Pending response rejected since connection got disposed
    at Object.dispose (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:51:4592)
    at Object.dispose (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:241:6209)
    at Nd.handleConnectionClosed (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:244:15575)
    at Nd.handleConnectionClosed (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:248:3561)
    at t (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:244:15224)
    at yT.invoke (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:14460)
    at lc.fire (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:15230)
    at Qe (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:25:6504)
    at yT.invoke (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:14460)
    at lc.fire (/Users/oepshtain/.vim/bundle/coc.nvim/build/index.js:24:15230) {
  code: [33m-32097[39m,
  data: [90mundefined[39m
}
2023-05-11T11:07:08.893 INFO (pid:45126) [attach] - receive notification: highlight []
2023-05-11T11:07:09.476 INFO (pid:45126) [attach] - receive notification: highlight []
2023-05-11T11:07:15.307 INFO (pid:45126) [attach] - receive notification: openLog []
```

**Comment by fannheyward** [comment]
clangd failed to start, coc.nvim as a client, couldn't communicate to the server. What's your clangd version?

**Comment by oren0e** [comment]
@fannheyward 
```
~ ❯ clangd --version                                                                                                                                                                                                                   
Homebrew clangd version 16.0.3
Features: mac+xpc
Platform: x86_64-apple-darwin20.6.0
```

**Comment by fannheyward** [comment]
@oren0e the `clangd` installed from GitHub is broken, you can delete `~/.config/coc/extensions/coc-clangd-data/`, coc-clangd can use `clangd` from your $PATH.

**Comment by oren0e** [comment]
@fannheyward I get the `clangd was not found on your PATH...` message, despite having this in my PATH (`/usr/local/opt/llvm/bin`):
```
~ ❯ which clangd                                                                                                                                                                                                                       
/usr/local/opt/llvm/bin/clangd
```

**Comment by fannheyward** [comment]
@oren0e set `clangd.path` to `/usr/local/opt/llvm/bin/clangd` in your coc-settings.json please.

**Comment by oren0e** [comment]
Thank you! that solved the problem, I should have done that, you're right. Cheers

**Comment by VasilySotnikov** [comment]
I get the same error after `clangd` installed from the command `CocCommand clangd.install` became version 16.0.2 a couple of days ago. If I switch back to 15.0.6 the problem disappears. 

Would be good to know if this is expected to be fixed later or one should not install clangd with `CocCommand` any more?

**Comment by fannheyward** [comment]
Looks like the clangd downloaded from GitHub is broken, will look into this.

**Comment by meinzer1899** [comment]
> I get the same error after `clangd` installed from the command `CocCommand clangd.install` became version 16.0.2 a couple of days ago. If I switch back to 15.0.6 the problem disappears.
> 
> Would be good to know if this is expected to be fixed later or one should not install clangd with `CocCommand` any more?

Same here

**Comment by fannheyward** [comment]
This has been fixed in v0.29.1, please `:CocUpdate`.

**Comment by meinzer1899** [comment]
Hey @fannheyward , thanks for the update.

I tried it, but was expecting a clangd error in `stdint.h`, provided via clangd in `clangd_16.0.2/lib/clang/16/include`.
For some reason, this file is broken, when installed (and unzipped) via `:CocCommand clangd.install`. If I unzip the archive as mentioned in https://github.com/clangd/coc-clangd/issues/577#issuecomment-1562607614, it is formatted:

![grafik](https://github.com/clangd/coc-clangd/assets/44314956/c032d4ba-8375-451c-9f29-3415b5299548)
(left: unzipped manually, right: installed via coc-clangd)

The header description 
> /*===---- stdint.h - Standard header for sized integer types --------------===*\

is moved to line 447 in the file installed via coc-clangd.
![grafik](https://github.com/clangd/coc-clangd/assets/44314956/c4a756d8-abcf-42e0-a7fa-108b18088894)
(left: unzipped manually, right: installed via coc-clangd).

Something seems to be wrong when unzipping, resulting in a(t least one) corrupt file.

Here, a similar issue was reported: https://github.com/ZJONSSON/node-unzipper/issues/271. I'm using node v18.16.1.

**Comment by fannheyward** [comment]
@meinzer1899 thank you for your report! Are you using node 18? Tested with node 20, the file is complete.

unzipper broken this.

**Comment by fannheyward** [comment]
@meinzer1899 tested with node v18.16.1 + unzipper@0.10.14, can't reproduce this. Can you delete `~/.config/coc/extensions/coc-clangd-data/` and try again?

```js
import * as unzipper from 'unzipper';
import * as fs from "fs";
import * as path from "path";

const p = '~/clangd-mac-16.0.2.zip';
const extraRoot = '/tmp/x1';
const archive = await unzipper.Open.file(p);
await archive.extract({ path: extraRoot });

const bin = path.join(extraRoot, 'clangd_16.0.2', 'bin', 'clangd');
fs.promises.chmod(bin, 0o755);
```

**Comment by fannheyward** [comment]
Try node 20.

**Comment by meinzer1899** [comment]
I have tried with node 20.4.0, the issue remains.
<details>
  <summary>Expand</summary>
![grafik](https://github.com/clangd/coc-clangd/assets/44314956/e3aae10e-51de-48f6-b89c-c59867b5aae5)
(left: unzipped manually, right: via clangd.install)

  
</details>

The reason why I previously had node 18.16.1 installed, is that its the LTS version. I think this will be the version to go for most (C++) developers, which want to use clangd.

Can we replace unzipper with an alternative, e.g. https://stackoverflow.com/questions/10308110/simplest-way-to-download-and-unzip-files-in-node-js-cross-platform?

**Comment by fannheyward** [comment]
@meinzer1899 https://github.com/clangd/node-clangd/pull/25

**Comment by fannheyward** [comment]
I've patched adm-zip to coc-clangd, don't need to wait node-clangd release, and published coc-clangd v0.29.3.
